### PR TITLE
Support closed-contract ECT registration reassignment to 2024 for partnership and EOI reuse

### PR DESCRIPTION
**Pull request by ericaporter**
### Context
Ticket: here

This PR implements the reassignment logic for ECTs whose latest training period is in a closed 2021 or 2022 contract period.

When such ECTs are registered again, their training is reassigned to the 2024 contract period, ensuring they can continue training under the appropriate contract terms.

### Changes proposed in this pull request

- Introduced ContractPeriods::ForECTRegistration to resolve the correct contract period for registration
   - Repurposed the previously unused ContractPeriods::ForECT into ContractPeriods::ForECTRegistration to explicitly handle contract period resolution during ECT registration flows
- Using ContractPeriods::Reassignment to encapsulate logic for moving ECTs from closed 2021/2022 provider-led periods to 2024
- Updated registration queries to pass previous_training_period into the resolver
- Ensured provider-led ECTs from closed 2021/2022 periods are reassigned to 2024
- Reuse of previous training choices now evaluated against the resolved (2024) contract period
- New training periods are created as:
   - confirmed partnership if available in 2024
   - expression of interest (EOI) if no partnership exists

Specs:
- Added coverage for 2021/2022 → 2024 reassignment, reuse (partnership/EOI), school view, and API behaviour
- Added unit tests for reuse logic and contract period resolution
- Refactored specs to reduce duplication, simplify setup, and group scenarios by reuse type

### Guidance to review

New Commits: 
Several edge cases in the reuse screen logic were found during testing and fixed:
    - EOI lookups fixed to scope to the current school and last chosen lead provider
    - payments_frozen_reassignment? now correctly handles EOI-based previous training periods 
    - Reuse screen now correctly shows for frozen contract period ECTs regardless of whether their previous training period used a confirmed partnership or EOI

### CODE REVIEW

**Review by cpjmcquillan — APPROVED** [review]

File: spec/features/schools/ects/register/closed_contract_period/closed_contract_period_spec.rb [hunk 1]
```diff
@@ -0,0 +1,352 @@
+RSpec.describe "Registering an ECT - closed contract period", :enable_schools_interface do
+  include_context "test TRS API returns a teacher"
+
+  around do |example|
+    travel_to(Date.new(2024, 9, 1)) { example.run }
+  end
```
[the remark this answers is not in the document]
Comment (ericaporter) [reply spec/features/schools/ects/register/closed_contract_period/closed_contract_period_spec.rb]: @cpjmcquillan I did a config check - I couldn't see a global time-reset hook. I think RSpec will clean up stubs between examples but that's separate to travel_to? If we switched to `before {travel_to...` etc. I think we would need to put an `after {travel_back...`

[the remark this answers is not in the document]
Comment (ericaporter) [reply spec/features/schools/ects/register/closed_contract_period/closed_contract_period_spec.rb]: Thanks for checking! - good to know this is being called automatically - will update to the before block

File: spec/features/schools/ects/register/closed_contract_period/closed_contract_period_spec.rb [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+    @previous_training_period = FactoryBot.create(
+      :training_period,
+      ect_at_school_period: @previous_ect_at_school_period,
+      training_programme: "provider_led",
+      started_on: Date.new(2022, 9, 1),
+      finished_on: Date.new(2023, 7, 31),
+      school_partnership: @previous_school_partnership,
+      expression_of_interest: nil
+    )
```
[the remark this answers is not in the document]
Comment (ericaporter) [reply spec/features/schools/ects/register/closed_contract_period/closed_contract_period_spec.rb]: Tried it again and it works - not sure what was going on either!

File: app/wizards/schools/register_ect_wizard/registration_store/queries.rb [hunk 3]
```diff
@@ -51,10 +54,17 @@ def lead_providers_within_contract_period
...
-            .new(order: :started_on)
-            .ect_periods(trn: registration_store.trn)
-            .last
+          @previous_ect_at_school_period ||= begin
+            periods = ECTAtSchoolPeriods::Search
+              .new(order: :started_on)
+              .ect_periods(trn: registration_store.trn)
+
+            if ect_at_school_period
+              periods = periods.where.not(id: ect_at_school_period.id)
```
Comment (cpjmcquillan) [inline app/wizards/schools/register_ect_wizard/registration_store/queries.rb]: There are a couple of shorthand methods for this in AR if you're interested 👌🏼

```suggestion
              periods = periods.without(ect_at_school_period)
```
or
```suggestion
              periods = periods.excluding(ect_at_school_period)
```

File: app/wizards/schools/register_ect_wizard/use_previous_ect_choices_step.rb [hunk 4]
```diff
@@ -129,11 +152,49 @@ def reusable_partnership_in_registration_year?
...
+          previous_eoi_training_period_for_reassignment(contract_period)
+        else
+          most_recent_school_eoi_training_period(contract_period)
+        end
+      end
+
+      def previous_eoi_training_period_for_reassignment(contract_period)
+        training_period = ect.previous_training_period
+        return unless training_period
+        return unless training_period.training_programme == "provider_led"
```
Comment (cpjmcquillan) [inline app/wizards/schools/register_ect_wizard/use_previous_ect_choices_step.rb]:
```suggestion
        return unless training_period.provider_led_training_programme?
```

File: spec/wizards/schools/register_ect_wizard/registration_store/queries_spec.rb [hunk 5]
```diff
@@ -99,53 +253,137 @@
...
+        create_active_lead_provider(contract_period: contract_period_2025, lead_provider: lead_provider_2025)
+        stub_previous_training(previous_training_period, previous_ect_period:)
+      end
+
+      it "returns lead providers for the 2024 contract period" do
+        names = queries.lead_providers_within_contract_period.map(&:name)
+
+        expect(names).to include("LP 2024")
+        expect(names).not_to include("LP 2025")
+      end
```
Comment (cpjmcquillan) [inline spec/wizards/schools/register_ect_wizard/registration_store/queries_spec.rb:285]: Do we want to use `contain_exactly` like the test below?

**Review by cpjmcquillan — APPROVED** [review]

File: app/wizards/schools/register_ect_wizard/registration_store/queries.rb (hunk 3 above)
Comment (ericaporter) [reply app/wizards/schools/register_ect_wizard/registration_store/queries.rb]: Yeah, great!
